File: iosApp/iosAppTests/feature_game/presentation/GameSceneHandlerTests.swift
//
//  GameSceneHandlerTests.swift
//  iosAppTests
//
//  Created by Bradley Phillips on 1/11/23.
//  Copyright © 2023 orgName. All rights reserved.
//

import XCTest
import ViewInspector
@testable import iosApp
@testable import shared
import SwiftUI

final class GameSceneHandlerTests: XCTestCase
{
    override func setUpWithError() throws
    {
        KoinPlatformManager.shared.stop()
        KoinPlatformManager.shared.start(
            scenarios: [
                Scenario.wordFound,
                Scenario.answerSaved
            ]
        )
    }
    
    override class func tearDown() { KoinPlatformManager.shared.stop() }
    
    func test_when_initialized__activeView_is_Empty()
    {
        // given
        let expectedActiveView = GameSceneHandler.ViewType.EMPTY
        
        // when
        let sut = GameSceneHandler()
        
        // then
        XCTAssertEqual(expectedActiveView, sut.activeView)
        XCTAssertFalse(sut.gameKeyboardIsEnabled)
    }
    
    func test_when_setUp_invoked__viewModel_setUp_is_invoked()
    {
        // given
        let expectedActiveView = GameSceneHandler.ViewType.GAME
        let sut = GameSceneHandler()
        
        // when
        sut.setUp()
        
        // then
        XCTAssertEqual(expectedActiveView, sut.activeView)
        XCTAssertTrue(sut.gameKeyboardIsEnabled)
    }
    
    func test_when_gameHeader_view_appears__title_matches_expected_title()
    {
        // given
        let expectedTitle = GameSceneViewModel().getHeader().title()
        let sut = GameSceneHandler()
        let actualHeader = sut.gameHeader()
        
        // when
        defer { ViewHosting.expel() }
        ViewHosting.host(view: actualHeader.environmentObject(SceneDimensions()))
        
        // then
        XCTAssertEqual(expectedTitle,try actualHeader.inspect().find(text: expectedTitle).string())
    }
    
    func test_when_gameBoard_view_appears__rows_match_expected_value()
    {
        // given
        let expectedRowCount = GameSceneViewModel().getGameBoard().rows().count
        let sut = GameSceneHandler()
        let actualBoard = sut.gameBoard()
        
        // when
        defer { ViewHosting.expel() }
        ViewHosting.host(view: actualBoard.environmentObject(SceneDimensions()))
        
        // then
        XCTAssertEqual(expectedRowCount, try actualBoard.inspect().vStack().forEach(0).count)
    }
    
    func test_when_gameKeyboard_view_appears__rows_match_expected_value()
    {
        // given
        let expectedRowCount = GameSceneViewModel().getGameKeyboard().rows().count
        let sut = GameSceneHandler()
        let actualKeyboard = sut.gameKeyboard()
        
        // when
        defer { ViewHosting.expel() }
        ViewHosting.host(view: actualKeyboard.environmentObject(SceneDimensions()))
        
        // then
        XCTAssertEqual(expectedRowCount, try actualKeyboard.inspect().zStack().vStack(0).forEach(0).count)
    }
    
    func test_when_gameAnnouncement_view_appears__text_matches_expected_value()
    {
        // given
        let expectedTextValue = "My Middle Layer Announcement"
        let actualAnnouncement = GameSceneViewModel().getAnnouncement()
        let sut = GameSceneHandler()
        actualAnnouncement.setMessage(newMessage: expectedTextValue)
        let actualAnnouncementView = sut.gameAnnouncement()
        
        // when
        defer { ViewHosting.expel() }
        ViewHosting.host(view: actualAnnouncementView.environmentObject(SceneDimensions()))
        
        // then
        XCTAssertEqual(expectedTextValue, try actualAnnouncementView.inspect().text().string())
    }
    
    func test_when_gameAnnouncement_contains_no_announcement__view_is_not_returned()
    {
        // given
        let actualAnnouncement = GameSceneViewModel().getAnnouncement()
        actualAnnouncement.setMessage(newMessage: nil)
        
        // when
        let actualAnnouncementView = GameSceneHandler().gameAnnouncement()
        
        // then
        XCTAssertNil(actualAnnouncementView)
    }
    
    func test_when_onGameOver_invoked__active_view_is_published_to_GAME()
    {
        // given
        let actualAnnouncement = GameSceneViewModel().getAnnouncement()
        let sut = GameSceneHandler()
        sut.onGameStarted()
        let expectation = XCTestExpectation(description: "Waiting for onGameOver to publish.")
        let mockView = MockView(handler: sut, expectation: expectation)
        defer { ViewHosting.expel() }
        ViewHosting.host(view: mockView)
        
        // when
        actualAnnouncement.setMessage(newMessage: "Game over, man!")
        sut.onGameOver()
        
        // then
        wait(for: [expectation], timeout: 1.0)
        XCTAssertEqual(.GAME, sut.activeView)
        XCTAssertFalse(sut.gameKeyboardIsEnabled)
        XCTAssertNoThrow(try mockView.inspect().find(text: "Game over, man!"))
    }
    
    func test_when_onRevealNextTile_invoked__active_view_is_published_to_GAME()
    {
        // given
        let actualAnnouncement = GameSceneViewModel().getAnnouncement()
        let sut = GameSceneHandler()
        sut.onGameStarted()
        let expectation = XCTestExpectation(description: "Waiting for onRevealNextTile to publish.")
        let mockView = MockView(handler: sut, expectation: expectation)
        defer { ViewHosting.expel() }
        ViewHosting.host(view: mockView)
        
        // when
        actualAnnouncement.setMessage(newMessage: "Revealing the next tile!")
        sut.onRevealNextTile()
        
        // then
        wait(for: [expectation], timeout: 1.0)
        XCTAssertEqual(.GAME, sut.activeView)
        XCTAssertTrue(sut.gameKeyboardIsEnabled)
        XCTAssertNoThrow(try mockView.inspect().find(text: "Revealing the next tile!"))
    }
    
    func test_when_onRoundCompleted_invoked__active_view_is_published_to_GAME()
    {
        // given
        let actualAnnouncement = GameSceneViewModel().getAnnouncement()
        let sut = GameSceneHandler()
        sut.onGameStarted()
        let expectation = XCTestExpectation(description: "Waiting for onRoundCompleted to publish.")
        let mockView = MockView(handler: sut, expectation: expectation)
        defer { ViewHosting.expel() }
        ViewHosting.host(view: mockView)
        
        // when
        actualAnnouncement.setMessage(newMessage: "Next Round!")
        sut.onRoundCompleted()
        
        // then
        wait(for: [expectation], timeout: 1.0)
        XCTAssertEqual(.GAME, sut.activeView)
        XCTAssertTrue(sut.gameKeyboardIsEnabled)
        XCTAssertNoThrow(try mockView.inspect().find(text: "Next Round!"))
    }
    
    func test_when_onStartingGame_invoked__active_view_is_published_to_GAME()
    {
        // given
        let actualAnnouncement = GameSceneViewModel().getAnnouncement()
        let sut = GameSceneHandler()
        sut.onGameStarted()
        let expectation = XCTestExpectation(description: "Waiting for onStartingGame to publish.")
        let mockView = MockView(handler: sut, expectation: expectation)
        defer { ViewHosting.expel() }
        ViewHosting.host(view: mockView)
        
        // when
        actualAnnouncement.setMessage(newMessage: "Start Game!")
        sut.onStartingGame()
        
        // then
        wait(for: [expectation], timeout: 1.0)
        XCTAssertEqual(.GAME, sut.activeView)
        XCTAssertFalse(sut.gameKeyboardIsEnabled)
        XCTAssertNoThrow(try mockView.inspect().find(text: "Start Game!"))
    }
    
    func test_when_onGuessingWord_invoked__active_view_is_published_to_GAME()
    {
        // given
        let actualAnnouncement = GameSceneViewModel().getAnnouncement()
        let sut = GameSceneHandler()
        sut.onGameStarted()
        let expectation = XCTestExpectation(description: "Waiting for onGuessingWord to publish.")
        let mockView = MockView(handler: sut, expectation: expectation)
        defer { ViewHosting.expel() }
        ViewHosting.host(view: mockView)
        
        // when
        actualAnnouncement.setMessage(newMessage: "Guessing word, hold your ponies!")
        sut.onGuessingWord()
        
        // then
        wait(for: [expectation], timeout: 1.0)
        XCTAssertEqual(.GAME, sut.activeView)
        XCTAssertFalse(sut.gameKeyboardIsEnabled)
        XCTAssertNoThrow(try mockView.inspect().find(text: "Guessing word, hold your ponies!"))
    }
    
    func test_when_gameKeyboard_is_enabled_and_key_clicked__expected_function_is_invoked() throws
    {
        // given
        let sut = GameSceneHandler()
        let keyboardView = sut.gameKeyboard()
        sut.setUp()
        
        defer { ViewHosting.expel() }
        ViewHosting.host(view: keyboardView.environmentObject(SceneDimensions()))
        
        // when
        sut.activeView = .EMPTY
        let key = try keyboardView.inspect().find(GameSceneKeyboard.Key.self, where: { try $0.actualView().letters == "K" } )
        try key.button().tap()
        
        // then
        XCTAssertEqual(.GAME, sut.activeView)
    }
    
    func test_when_gameKeyboard_is_not_enabled_and_key_clicked__no_function_is_invoked() throws
    {
        // given
        let sut = GameSceneHandler()
        let keyboardView = sut.gameKeyboard()
        sut.setUp()
        sut.onGuessingWord()
        
        defer { ViewHosting.expel() }
        ViewHosting.host(view: keyboardView.environmentObject(SceneDimensions()))
        
        // when
        sut.activeView = .EMPTY
        let key = try keyboardView.inspect().find(GameSceneKeyboard.Key.self, where: { try $0.actualView().letters == "K" } )
        try key.button().tap()
        
        // then
        XCTAssertEqual(.EMPTY, sut.activeView)
    }
    
    struct MockView: View
    {
        @ObservedObject var handler: GameSceneHandler
        
        var expectation: XCTestExpectation
        
        var body: some View
        {
            Text(handler.gameAnnouncement()?.text ?? "")
                .onReceive(handler.$activeView)
            { newActiveView in
                if newActiveView != .EMPTY { expectation.fulfill() }
            }
        }
    }
}

File: androidApp/src/androidTest/java/com/megabreezy/breezybuilds_wordle/game/presentation/GameSceneHandlerTests.kt
package com.megabreezy.breezybuilds_wordle.game.presentation

import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.test.*
import androidx.compose.ui.test.junit4.createComposeRule
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.megabreezy.breezybuilds_wordle.android.game.presentation.GameSceneHandler
import com.megabreezy.breezybuilds_wordle.android.game.presentation.component.GameSceneAnnouncement
import com.megabreezy.breezybuilds_wordle.android.game.presentation.component.GameSceneBoard
import com.megabreezy.breezybuilds_wordle.android.game.presentation.component.GameSceneHeader
import com.megabreezy.breezybuilds_wordle.android.game.presentation.component.GameSceneKeyboard
import com.megabreezy.breezybuilds_wordle.android.game.presentation.rememberGameSceneHandler
import com.megabreezy.breezybuilds_wordle.core.ui.SceneMock
import com.megabreezy.breezybuilds_wordle.core.util.KoinPlatformManager
import com.megabreezy.breezybuilds_wordle.core.util.Scenario
import com.megabreezy.breezybuilds_wordle.feature.game.presentation.GameSceneViewModel
import kotlinx.coroutines.runBlocking
import org.junit.*
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class GameSceneHandlerTests
{
    @get:Rule
    val composeTestRule = createComposeRule()

    @Before
    fun setUp()
    {
        KoinPlatformManager.start(
            scenarios = listOf(
                Scenario.WORD_FOUND,
                Scenario.ANSWER_SAVED
            )
        )
    }

    @Test
    fun when_initialized__activeView_is_Empty()
    {
        // given
        val expectedActiveView = GameSceneHandler.ViewType.EMPTY

        // when
        val sut = GameSceneHandler()

        // then
        Assert.assertEquals(expectedActiveView, sut.activeView)
        Assert.assertFalse(sut.gameKeyboardIsEnabled)
    }

    @Test
    fun when_setUp_invoked__viewModel_setUp_is_invoked()
    {
        // given
        val expectedActiveView = GameSceneHandler.ViewType.GAME
        val sut = GameSceneHandler()

        // when
        runBlocking { sut.setUp() }

        // then
        Assert.assertEquals(expectedActiveView, sut.activeView)
        Assert.assertTrue(sut.gameKeyboardIsEnabled)
    }

    @Test
    fun when_GameHeader_view_appears__title_matches_expected_title()
    {
        // given
        val expectedTitle = GameSceneViewModel().getHeader().title()
        val sut = GameSceneHandler()

        // when
        composeTestRule.setContent()
        {
            SceneMock.display { sut.GameHeader() }
        }

        // then
        composeTestRule.onNodeWithContentDescription(
            GameSceneHeader.TagName.COMPONENT.toString()
        ).onChild().assertTextEquals(expectedTitle)
    }

    @Test
    fun when_GameBoard_view_appears__rows_match_expected_value()
    {
        // given
        lateinit var handler: GameSceneHandler
        val expectedRowCount = runBlocking { GameSceneViewModel().getGameBoard().rows().count() }
        val expectedTilesCount = runBlocking { expectedRowCount * (GameSceneViewModel().getGameBoard().rows().last().count()) }

        // when
        composeTestRule.setContent()
        {
            handler = rememberGameSceneHandler(scope = rememberCoroutineScope())

            SceneMock.display { handler.GameBoard() }
        }

        runBlocking()
        {
            handler.setUp()
        }

        // then
        composeTestRule.onNodeWithContentDescription(
            GameSceneBoard.TagName.BOARD.toString(), useUnmergedTree = true
        ).onChildAt(index = expectedTilesCount - 1).assertExists()
    }

    @Test
    fun when_gameKeyboard_view_appears__rows_match_expected_value()
    {
        // given
        lateinit var handler: GameSceneHandler
        var expectedKeyCount = 0
        GameSceneViewModel().getGameKeyboard().rows().forEach { expectedKeyCount += it.count() }

        // when
        composeTestRule.setContent()
        {
            handler = rememberGameSceneHandler()

            SceneMock.display { handler.GameKeyboard() }
        }

        runBlocking { handler.setUp() }

        // then
        composeTestRule.onNodeWithContentDescription(
            GameSceneKeyboard.TagName.ROWS.toString(), useUnmergedTree = true
        ).onChildAt(index = expectedKeyCount - 1).assertExists()
    }

    @Test
    fun when_GameAnnouncement_view_appears__text_matches_expected_value()
    {
        // given
        lateinit var handler: GameSceneHandler
        val expectedTextValue = "My Middle Layer Announcement"
        val actualAnnouncement = GameSceneViewModel().getAnnouncement()
        actualAnnouncement.setMessage(newMessage = expectedTextValue)

        // when
        composeTestRule.setContent()
        {
            handler = rememberGameSceneHandler()

            SceneMock.display { handler.GameAnnouncement() }
        }

        // then
        composeTestRule.onNodeWithContentDescription(
            GameSceneAnnouncement.TagName.COMPONENT.toString(), useUnmergedTree = true
        ).assertTextEquals(expectedTextValue)
    }

    @Test
    fun when_onRevealNextTile_invoked__active_view_is_published_to_GAME()
    {
        // given
        lateinit var handler: GameSceneHandler
        val actualAnnouncement = GameSceneViewModel().getAnnouncement()

        // when
        composeTestRule.setContent()
        {
            handler = rememberGameSceneHandler()

            LaunchedEffect(Unit) { handler.setUp() }

            SceneMock.display { handler.GameAnnouncement() }
        }
        actualAnnouncement.setMessage(newMessage = "Revealing next tile!")
        handler.onRevealNextTile()

        // then
        Assert.assertEquals(GameSceneHandler.ViewType.GAME, handler.activeView)
        Assert.assertTrue(handler.gameKeyboardIsEnabled)
        composeTestRule.onNodeWithContentDescription(
            GameSceneAnnouncement.TagName.COMPONENT.toString(), useUnmergedTree = true
        ).assertTextEquals("Revealing next tile!")
    }

    @Test
    fun when_onRoundCompleted_invoked__active_view_is_published_to_GAME()
    {
        // given
        lateinit var handler: GameSceneHandler
        val actualAnnouncement = GameSceneViewModel().getAnnouncement()

        // when
        composeTestRule.setContent()
        {
            handler = rememberGameSceneHandler()

            LaunchedEffect(Unit) { handler.setUp() }

            SceneMock.display { handler.GameAnnouncement() }
        }
        actualAnnouncement.setMessage(newMessage = "Next Round!")
        handler.activeView = GameSceneHandler.ViewType.EMPTY
        handler.onRoundCompleted()

        // then
        Assert.assertEquals(GameSceneHandler.ViewType.GAME, handler.activeView)
        Assert.assertTrue(handler.gameKeyboardIsEnabled)
        composeTestRule.onNodeWithContentDescription(
            GameSceneAnnouncement.TagName.COMPONENT.toString(), useUnmergedTree = true
        ).assertTextEquals("Next Round!")
    }

    @Test
    fun when_onStartingGame_invoked__active_view_is_published_to_GAME()
    {
        // given
        lateinit var handler: GameSceneHandler
        val actualAnnouncement = GameSceneViewModel().getAnnouncement()

        // when
        composeTestRule.setContent()
        {
            handler = rememberGameSceneHandler()

            LaunchedEffect(Unit) { handler.setUp() }

            SceneMock.display { handler.GameAnnouncement() }
        }
        actualAnnouncement.setMessage(newMessage = "Start Game!")
        handler.activeView = GameSceneHandler.ViewType.EMPTY
        handler.onStartingGame()

        // then
        Assert.assertEquals(GameSceneHandler.ViewType.GAME, handler.activeView)
        Assert.assertFalse(handler.gameKeyboardIsEnabled)
        composeTestRule.onNodeWithContentDescription(
            GameSceneAnnouncement.TagName.COMPONENT.toString(), useUnmergedTree = true
        ).assertTextEquals("Start Game!")
    }

    @Test
    fun when_onGuessingWord_invoked__active_view_is_published_to_GAME()
    {
        // given
        lateinit var handler: GameSceneHandler
        val actualAnnouncement = GameSceneViewModel().getAnnouncement()

        // when
        composeTestRule.setContent()
        {
            handler = rememberGameSceneHandler()

            LaunchedEffect(Unit) { handler.setUp() }

            SceneMock.display { handler.GameAnnouncement() }
        }
        actualAnnouncement.setMessage(newMessage = "Guessing word, hold your ponies!")
        handler.activeView = GameSceneHandler.ViewType.EMPTY
        handler.onGuessingWord()

        // then
        Assert.assertEquals(GameSceneHandler.ViewType.GAME, handler.activeView)
        Assert.assertFalse(handler.gameKeyboardIsEnabled)
        composeTestRule.onNodeWithContentDescription(
            GameSceneAnnouncement.TagName.COMPONENT.toString(), useUnmergedTree = true
        ).assertTextEquals("Guessing word, hold your ponies!")
    }

    @Test
    fun when_onAnnouncementShouldShow_invoked__active_view_is_published_to_GAME()
    {
        // given
        lateinit var handler: GameSceneHandler
        val actualAnnouncement = GameSceneViewModel().getAnnouncement()

        // when
        composeTestRule.setContent()
        {
            handler = rememberGameSceneHandler()

            LaunchedEffect(Unit) { handler.setUp() }

            SceneMock.display { handler.GameAnnouncement() }
        }
        actualAnnouncement.setMessage(newMessage = "Not in words list!")
        handler.activeView = GameSceneHandler.ViewType.EMPTY
        handler.onAnnouncementShouldShow()

        // then
        Assert.assertEquals(GameSceneHandler.ViewType.GAME, handler.activeView)
        Assert.assertTrue(handler.gameKeyboardIsEnabled)
        composeTestRule.onNodeWithContentDescription(
            GameSceneAnnouncement.TagName.COMPONENT.toString(), useUnmergedTree = true
        ).assertTextEquals("Not in words list!")
    }

    @Test
    fun when_onAnnouncementShouldHide_invoked__active_view_is_published_to_GAME()
    {
        // given
        lateinit var handler: GameSceneHandler
        val actualAnnouncement = GameSceneViewModel().getAnnouncement()
        actualAnnouncement.setMessage(newMessage = "This message should be hidden!")

        // when
        composeTestRule.setContent()
        {
            handler = rememberGameSceneHandler()

            LaunchedEffect(Unit) { handler.setUp() }

            SceneMock.display { handler.GameAnnouncement() }
        }
        actualAnnouncement.setMessage(newMessage = null)
        handler.activeView = GameSceneHandler.ViewType.EMPTY
        handler.onAnnouncementShouldHide()

        // then
        Assert.assertEquals(GameSceneHandler.ViewType.GAME, handler.activeView)
        Assert.assertTrue(handler.gameKeyboardIsEnabled)
        composeTestRule.onNodeWithContentDescription(
            GameSceneAnnouncement.TagName.COMPONENT.toString(), useUnmergedTree = true
        ).assertDoesNotExist()
    }

    @Test
    fun when_onGameOver_invoked__active_view_is_published_to_GAME()
    {
        // given
        lateinit var handler: GameSceneHandler
        val actualAnnouncement = GameSceneViewModel().getAnnouncement()

        // when
        composeTestRule.setContent()
        {
            handler = rememberGameSceneHandler()

            LaunchedEffect(Unit) { handler.setUp() }

            SceneMock.display { handler.GameAnnouncement() }
        }
        actualAnnouncement.setMessage(newMessage = "Game over, man!")
        handler.activeView = GameSceneHandler.ViewType.EMPTY
        handler.onGameOver()

        // then
        Assert.assertEquals(GameSceneHandler.ViewType.GAME, handler.activeView)
        Assert.assertFalse(handler.gameKeyboardIsEnabled)
        composeTestRule.onNodeWithContentDescription(
            GameSceneAnnouncement.TagName.COMPONENT.toString(), useUnmergedTree = true
        ).assertTextEquals("Game over, man!")
    }

    @Test
    fun when_gameKeyboard_is_enabled_and_key_clicked__expected_function_is_invoked()
    {
        // given
        lateinit var handler: GameSceneHandler

        // when
        composeTestRule.setContent()
        {
            handler = rememberGameSceneHandler(scope = rememberCoroutineScope())

            LaunchedEffect(Unit) { handler.setUp() }

            SceneMock.display { handler.GameKeyboard() }
        }
        runBlocking { handler.setUp() }
        handler.activeView = GameSceneHandler.ViewType.EMPTY
        composeTestRule.onAllNodesWithContentDescription(
            GameSceneKeyboard.TagName.KEY.toString()
        ).onFirst().performClick()

        // then
        Assert.assertEquals(GameSceneHandler.ViewType.GAME, handler.activeView)
    }

    @Test
    fun when_gameKeyboard_is_not_enabled_and_key_clicked__no_function_is_invoked()
    {
        // given
        lateinit var handler: GameSceneHandler

        // when
        composeTestRule.setContent()
        {
            handler = rememberGameSceneHandler()

            LaunchedEffect(Unit) { handler.setUp() }

            SceneMock.display { handler.GameKeyboard() }
        }
        runBlocking { handler.setUp() }
        handler.onGuessingWord()
        handler.activeView = GameSceneHandler.ViewType.EMPTY
        composeTestRule.onAllNodesWithContentDescription(
            GameSceneKeyboard.TagName.KEY.toString()
        ).onFirst().performClick()

        // then
        Assert.assertEquals(GameSceneHandler.ViewType.EMPTY, handler.activeView)
    }
}